plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,198 @@
+        this.bindDistinguishedName = ldapConfig.getBindDistinguishedName();
+        this.bindPassword = ldapConfig.getBindPassword();
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+        this.groupSearchFilter = ldapConfig.getGroupSearchFilter();
+        this.groupCache = buildNonEvictableCacheWithWeakInvalidateAll(
+                CacheBuilder.newBuilder()
+                        .expireAfterWrite(ldapConfig.getLdapCacheTtl().toMillis(), MILLISECONDS),
+                CacheLoader.from(this::getGroupsWithBindDistinguishedName));
+    }
+
+    @VisibleForTesting
+    void invalidateCache()
+    {
+        groupCache.invalidateAll();
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            return groupCache.getUnchecked(user);
+        }
+        catch (UncheckedExecutionException e) {
+            throw e;
+        }
+    }
+
+    private Set<String> getGroupsWithBindDistinguishedName(String user)
+    {
+        Set<String> groupSet = Collections.emptySet();
+
+        if (containsSpecialCharacters(user)) {
+            throw new AccessDeniedException("Username contains a special LDAP character");
+        }
+
+        try {
+            for (String userBindSearchPattern : userBindSearchPatterns) {
+                String searchFilter = replaceUser(userBindSearchPattern, user);
+                String userDistinguishedName = lookupUserDistinguishedName(
+                        this.userBaseDistinguishedName.orElseThrow(),
+                        searchFilter,
+                        this.bindDistinguishedName,
+                        this.bindPassword);
+                String groupFilter = replaceUser(this.groupSearchFilter, userDistinguishedName);
+                log.debug("Querying LDAP from %s using filter %s", this.groupBaseDistinguishedName, groupFilter);
+                Set<String> queryResult = lookupUserDistinguishedNames(
+                        this.groupBaseDistinguishedName, groupFilter,
+                        bindDistinguishedName, bindPassword);
+                groupSet = queryResult.stream().map(source -> {
+                    try {
+                        log.debug("extracting group from %s", source);
+                        Matcher m = EXTRACT_GROUP_PATTERN.matcher(source);
+                        if (m.find()) {
+                            return m.group(1);
+                        }
+                        log.debug("Cannot extract group cn [%s]", source);
+                        return "";
+                    }
+                    catch (Exception e) {
+                        log.debug(e, "Unknown exception, %s", e.getMessage());
+                        return "";
+                    }
+                }).filter(Objects::nonNull).collect(Collectors.toSet());
```

Comment:
Can we use `guava` Immutable collections.

Reply:
core java17 now has Collectors.toUnmodifiableSet() which achieves the same thing?, will put this in here